File: python/cat.py
#!/usr/bin/env python3

import numpy as np

def readseq(filename):
    label, seq = [i.strip() for i in open(filename).readlines()]
    return seq

def correct_subgraphs(seq):
    pass

def catalan(n):
    n = int(n)
    if n <= 1:
        mem = np.zeros((2, 2))
    else:
        mem = np.zeros((n + 1, n + 1))
    mem[0, 1], mem[1, 1] = 1, 1
    cat = lambda i: sum(mem[i])
    for i in range(2, n + 1):
        for j in range(1, i + 1):
            mem[i, j] = cat(j - 1)*cat(i - j)
    return cat(n)

if __name__ == '__main__':
    seq = readseq('test.txt')
    correct_subgraphs(seq)
    print(catalan(4))

File: julia/cat.jl
#!/usr/bin/env julia

type Graph
    nodes::Array
    edges::Array{Tuple}
end

function Graph(seq::String)
    nodes = [i for i in 1:length(seq)]
    edges = [(i, j) for i in nodes for j in nodes if i < j]
    return Graph(nodes, edges)
end

function readseq(filename)
    label, seq = [strip(line) for line in open(readlines, filename)]
    @assert count(i->i == 'A', seq) == count(i->i == 'U', seq)
    @assert count(i->i == 'G', seq) == count(i->i == 'C', seq)
    return seq
end

function ncpmatchings(graph::Graph)
    ncp = []
    for i in graph.nodes
        for j in graph.nodes[2i:end]
            println(i, ' ', j)
        end
    end
end

################################################################################
## Catalan Numbers
################################################################################

function rcatalan(n::Int)
    if n <= 1
        return 1
    else
        return sum(rcatalan(k-1)*rcatalan(n-k) for k in 1:n)
    end
end

function catalan(n::Int)
    mem = zeros(Int, n+1, n)
    mem[1, 1] = 1
    mem[2, 1] = 1
    for i in 3:n+1
        for k in 1:n
            mem[i, k] = sum(mem[k-1, 1:end])*sum(mem[n-k, 1:end])
        end
    end
    return mem
end

function main()
    mem = catalan(4)
    print(join(split(string(mem), ';'), '\n'))
end

main()
